libyul/optimiser/UnusedStoreEliminator.cpp

```diff
@@ -105,10 +105,13 @@ void UnusedStoreEliminator::operator()(FunctionCall const& _functionCall)
 	else
 		sideEffects = m_controlFlowSideEffects.at(_functionCall.functionName.name);
 
+	if (sideEffects.canTerminate)
+		changeUndecidedTo(State::Used, Location::Storage);
```

Comment:
So this now catches the `canContinue && canTerminate` case that was not handled before. I'm not sure I understand why in this case we were ending up with unused. I see in the file that there are only two situations in which we mark a storage write unused:
1. The `!sideEffects.canTerminate` case below.
2. `knownCovered()` in `applyOperation()`.

And I don't think that it was the second.

Were we just leaving it undecided until we encounter another function call? But then what if that's the only function call?

Reply:
It just remained undecided. Let me track the state of the write ``x = 1`` in the following, maybe that helps:
```
x = 1;
// undecided
f();
// still undecided (buggy! since f() may successfully terminate,
// it should already be marked as *used* after this call!)
x = 2;
// Now it changed from undecided to unused - the value was overwritten,
// we don't need the write above. That's of course wrong if ``f()`` can terminate,
// but that's what was missing - we assume that control flow *always* continues after ``f()``.
...
function f() { if (<sth>) { return; } assembly { return(0,0) } }
```
So the problem is that the call to ``f`` *doesn't* fix the write as "used", but leaves it undecided - not that it itself marks it as "unused". If the write is left undecided, a subsequent write may overwrite it (``knownCovered``) and change it to ``unused``, even though it should already have been fixed as used.

Reply:
Probably the same will happen with
```
x = 1;
// undecided
f();
// still undecided
revert();
// unused, since we assume the transaction always reverts anyways
```

That may actually even be the more common and more dangerous case - we'll need to check that. But yeah, then the ``!canContinue && !canTerminate`` should kick in and mark it as unused, once we hit the ``revert();``

Reply:
Thanks.

I was worried about the `canContinue && canRevert` case, but looks like that's not a problem after all. On the other hand yours with `revert()` does look bad.